### PR TITLE
Feature/cdap 2840 pfs upgrade

### PR DESCRIPTION
**Pull request by anwar6953**
PartitionedFileSet's partitions table used to be `Table`. It is now an `IndexedTable`, indexed on two fields: creation time and transaction write pointer. Both of these fields are columns on the rows that did not exist before 3.1.0.

https://issues.cask.co/browse/CDAP-2840
http://builds.cask.co/browse/CDAP-RBT311-3

### CODE REVIEW

**Comment by anwar6953** [comment]
This needs to be updated to also work for TimePartitionedFileSets

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java [hunk 1]
```diff
@@ -100,7 +100,7 @@
   /**
    * Column key used to store the existence of a row in the secondary index.
    */
-  private static final byte[] IDX_COL = {'r'};
+  public static final byte[] IDX_COL = {'r'};
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java]: does it have to be public? Since we are in the API, I am not sure we want to make this public.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+   * @param namespaceId the namespace that contains the PartitionedFileSet
+   * @param dsSpec the specification of the PartitionedFileSet that needs migrating
+   * @throws Exception
+   */
+  public void upgrade(Id.Namespace namespaceId, DatasetSpecification dsSpec) throws Exception {
+    DatasetSpecification newPartitionsSpec = dsSpec.getSpecification(PartitionedFileSetDefinition.PARTITION_TABLE_NAME);
+    TableId oldPartitionsTableId = TableId.from(namespaceId, newPartitionsSpec.getName());
+    TableId indexTableId = TableId.from(namespaceId, newPartitionsSpec.getName() + ".i");
+    TableId dataTableId = TableId.from(namespaceId, newPartitionsSpec.getName() + ".d");
+
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: remove extra newline

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    HTable oldTable = tableUtil.createHTable(conf, oldPartitionsTableId);
+    HTable newIndexTable = tableUtil.createHTable(conf, indexTableId);
+    HTable newDataTable = tableUtil.createHTable(conf, dataTableId);
+
+    LOG.info("Starting upgrade for table {}", Bytes.toString(oldTable.getTableName()));
+    try {
+      ScanBuilder scan = tableUtil.buildScan();
+      scan.setTimeRange(0, HConstants.LATEST_TIMESTAMP);
+      scan.setMaxVersions(1); // we only need to see one version of each row
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: Not true. What if a partition was created by a MR job, and that job then failed? The transaction would be invalidated (and not rolled back). That is, the write remains, but its write version is now in the invalid list maintained by the tx manager. So, you need to talk to tx manager, get a transaction. Count the number of excludes in that transaction, and that is the max version that you need. Then, while reading, you need to filter out excluded versions.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+          Long hBaseVersion = null;
+
+          for (Map.Entry<byte[], NavigableMap<byte[], NavigableMap<Long, byte[]>>> familyMap :
+            result.getMap().entrySet()) {
+            for (Map.Entry<byte[], NavigableMap<Long, byte[]>> columnMap : familyMap.getValue().entrySet()) {
+              for (Map.Entry<Long, byte[]> columnEntry : columnMap.getValue().entrySet()) {
+                Long timeStamp = columnEntry.getKey();
+                byte[] colVal = columnEntry.getValue();
+                put.add(familyMap.getKey(), columnMap.getKey(), timeStamp, colVal);
+                LOG.info("Migrating row. family: '{}', column: '{}', ts: '{}', colVal: '{}'",
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: that would create a lot of log messages. Make it debug?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+          // here, we make the assumption that dropping the six right-most digits of the transaction write pointer
+          // yields the timestamp at which it started
+          byte[] creationTimeBytes = Bytes.toBytes(hBaseVersion / MILLION);
+          put.add(columnFamily, PartitionedFileSetDataset.CREATION_TIME_COL,
+                  hBaseVersion, creationTimeBytes);
+
+          newDataTable.put(put);
+
+          // index the data table on the two columns
+          Put indexTableWritePointer = new Put(IndexedTable.createIndexKey(result.getRow(),
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: call this variable indexPut instead of indexTableWritePointer? it is a put, not a write pointer...

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+      ScanBuilder scan = tableUtil.buildScan();
+      scan.setTimeRange(0, HConstants.LATEST_TIMESTAMP);
+      scan.setMaxVersions(1); // we only need to see one version of each row
+      try (ResultScanner resultScanner = oldTable.getScanner(scan.build())) {
+        Result result;
+        while ((result = resultScanner.next()) != null) {
+          Put put = new Put(result.getRow());
+
+
+          Long hBaseVersion = null;
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: why not call this the writeVersion?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+      } finally {
+        oldTable.close();
+        newIndexTable.close();
+        newDataTable.close();
+      }
+      LOG.info("Successfully migrated data from table {} Now deleting it.", Bytes.toString(oldTable.getTableName()));
+      hBaseAdmin.deleteTable(oldTable.getTableName());
+      LOG.info("Succsefully deleted old data table {}", Bytes.toString(oldTable.getTableName()));
+    } catch (Exception e) {
+      LOG.error("Error while migrating data from table: {}", oldPartitionsTableId, e);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: In case of an error: Say I have migrated 1 out of 3 pfs's, and the 2nd one fails for some reason. The user will have to run the upgrade tool again. Will this method now fail for #1? And with it the upgrade tool?

File: cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractQueueUpgrader.java [hunk 8]
```diff
@@ -104,8 +104,8 @@ void upgrade() throws Exception {
               LOG.debug("Adding entry {} -> {} for upgrade",
                         Bytes.toString(entry.getKey()), Bytes.toString(entry.getValue()));
               put.add(QueueEntryRow.COLUMN_FAMILY, entry.getKey(), entry.getValue());
-              mutations.add(put);
             }
+            mutations.add(put);
```
Comment (anew) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractQueueUpgrader.java]: unrelated change?

File: cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+  boolean alreadyUpgraded(DatasetSpecification dsSpec) {
+    DatasetSpecification partitionsSpec = dsSpec.getSpecification(PartitionedFileSetDefinition.PARTITION_TABLE_NAME);
+    // the partitions table is now an IndexedTable (was Table in < v3.1.0 CDAP)
+    return isIndexedTable(partitionsSpec);
+  }
+
+  boolean isPartitionedFileSet(DatasetSpecification dsSpec) {
+    String dsType = dsSpec.getType();
+    return (PartitionedFileSet.class.getName().equals(dsType) || "partitionedFileSet".equals(dsType));
```
Comment (anew) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java]: add support for timePartitionedFileSet?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java (hunk 3 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: Hmm, a simpler method would to simple get all versions, and write them using the same versions as in the original table, right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java (hunk 7 above)
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: Never mind. I see that is dealt with in the PFSUpgrader

Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: Actually, this can happen. In the very rare case that a dataset embeds two PFS, the PFSUpgrader first rewrites the spec and stores it in MDS, then attempts to migrate both embedded datasets. If the second one fails, then the user retries the upgrade, it will not even attempt to migrate the embedded tables again.

Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: oh and you check for that before opening it. I think it is correct the way it is.

**Comment by anew** [comment]
@anwar6953 please see my comments.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java (hunk 7 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: Hmm, the only thing I still need to do is delete the old table, once it's done migrating (which isn't absolutely necessary, though).

File: cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractQueueUpgrader.java (hunk 8 above)
Comment (anwar6953) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractQueueUpgrader.java]: yea, AbstractQueueUpgrader was adding the Put to the mutations before the put was fully built.
Changed it to only add it to the mutation, once all changes to the put have been mode (After the for loop).

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java (hunk 1 above)
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java]: It seems difficult to not make this public, since the migrator class is in the same package is `PartitionedFileSet`, in order to make the variables we need in `PartitionedFileSet` not public, but just protected.
If we instantiated the indexed table and did transactional writes to it, that would be one way to get round it.
Alternatively, we could at least document that it shouldn't be used?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java (hunk 3 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: https://github.com/caskdata/cdap/commit/8e6cf081518516d5ed40ea81a965f54a714a073e

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java (hunk 7 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: Nevermind, it was already deleting the table. Changed it to use tableUtil's dropTable though, since that disables+deletes.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java (hunk 2 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: done

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java (hunk 1 above)
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java]: Since this will only be done one time (3.0 -> 3.1) and we will delete this code after 3.1, I almost think it is better to copy it.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java (hunk 3 above)
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java]: yeah and that is ok, because in 99%+ of the cases there will only be one version.

**Comment by anew** [comment]
Looks good except that I think the IndexedTable constants should not be made public.

**Comment by anew** [comment]
Oh, and add a test case for TPFS?

**Comment by anwar6953** [comment]
Added tpfs spec upgrade.
Once code review is complete, I'll test on a cluster once again (due to code refactoring).

**Comment by anew** [comment]
LGTM. Please test on a cluster before merging.

**Comment by anew** [comment]
can this be merged?

**Comment by anwar6953** [comment]
I have yet to test this again after the refactoring. Will do that today or early tomorrow and merge following that.

**Comment by anwar6953** [comment]
@andreas pushed one last commit, which resolves an issue which I discovered while testing the upgrade.
Other than that, this is good to merge.

**Comment by anew** [comment]
Good catch. I could have found that in the first round of review... 
LGTM

**Comment by anwar6953** [comment]
Squashed commits; will merge once build passes.
